pkg/apis/duck/v1alpha1/annotations.go

```diff
@@ -161,3 +169,25 @@ func validateAnnotationNotExists(annotations map[string]string, annotation strin
 	}
 	return errs
 }
+
+// ValidateClusterNameAnnotation validates the cluster-name annotation.
+func ValidateClusterNameAnnotation(annotations map[string]string, errs *apis.FieldError, client metadataClient.Client) *apis.FieldError {
+	if client.OnGCE() {
+		if _, err := utils.ClusterName(annotations[ClusterNameAnnotation], client); err != nil {
+			// If metadata access is disabled for some reason, validation will fail and ask user to provide clusterName by adding this annotation.
+			return errs.Also(apis.ErrGeneric(
+				fmt.Sprintf(`not able to get cluster name with error: %s, please provide it by adding annotation "cluster-name=$CLUSTER_NAME"`, err.Error()),
+				fmt.Sprintf("metadata.annotations[%s]", ClusterNameAnnotation)))
+		}
+	}
+	return errs
+}
+
+func SetClusterNameAnnotation(obj *metav1.ObjectMeta, client metadataClient.Client) {
+	if client.OnGCE() {
```

Comment:
If the annotation is already present, this method does nothing. Similarly, in the Validate method above, if the annotation is present we always return success. I think this is the right behavior.

Instead of checking for OnGCE first, it seems like we should check for the annotation first. Then we can avoid calling the metadata service most of the time.

This is optional and can be ignored or done in a followup.